Dev

awertynaxis opened the pull request:
refactoried views,serializers,urls

Commit:
Created 3 apps : user, item, trading. Created and presented all models. Writed serializers for all views. Added Dockerfile and dockerignore

Commit:
Refactoring all views, serializers,urls in apps,
writed some custom permissions, made fixtures, added .env file, added related name on models where its wanst done befor

Commit:
Added celery and task, made jwt auth with token

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -11,16 +11,20 @@
+import environ
+import os
+
+env = environ.Env()
```
As usual in any kind of project developers use python-dotenv package instead of django-environ

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -68,16 +80,48 @@
+        'USER': env('DATABASE_USER'),
+        'PASSWORD': env('DATABASE_PASSWORD'),
+        'HOST': env('DATABASE_HOST'),
+        'PORT': env('DATABASE_PORT'),
```
it's better to have default values when you are getting environment variables

KittenAstronaut commented on item/fixtures/currency.json:
```diff
@@ -0,0 +1,65 @@
+        "is_not_deleted": true
+    }
+},
+{
+    "model": "item.currency",
+    "pk": 4,
+    "fields": {
+        "code": "GBP",
+        "name": "Pound",
+        "is_not_deleted": true
+    }
+},
+{
+    "model": "item.currency",
+    "pk": 5,
+    "fields": {
+        "code": "RU",
+        "name": "Russian rubles",
+        "is_not_deleted": true
+    }
+},
+{
+    "model": "item.currency",
+    "pk": 6,
+    "fields": {
+        "code": "Huy",
+        "name": "Huylo",
+        "is_not_deleted": true
+    }
+},
+{
+    "model": "item.currency",
+    "pk": 7,
+    "fields": {
+        "code": "JO",
+        "name": "Jooo",
+        "is_not_deleted": true
+    }
+}
+]
```
What is this?

KittenAstronaut commented on item/managers.py:
```diff
@@ -0,0 +1,7 @@
+
+class CurrencyManager(models.Manager):
+    def is_not_deleted(self) -> QuerySet:
+        return self.filter(is_not_deleted=True)
```
It's better to call `filter_not_deleted` instead of `is_not_deleted` because of naming convention
variables and functions that starts from `is_...` as usual are boolean or returns bool.

KittenAstronaut replied:
and also it's better to name fields `is_...` instead of `is_not_...` it's easier to read and filter

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,95 @@
+        if value == 'Huy':
+            raise serializers.ValidationError('Sorry u cant use Huy'
+                                              ' for currency naming')
+        return value
```
why are they static

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,95 @@
+    @staticmethod
+    def validate_code(value):
+
+        if value not in value.upper():
```
why not `value == value.upper()`?

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,95 @@
+class ItemDetailSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Item
+        fields = '__all__'
```
do not use `__all__` in field it's better to list all fields.
Because if we extend our Model and there's no needed field for this serializer it will include it

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,95 @@
+class PriceSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Price
+        fields = '__all__'
```
Same

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,95 @@
+class PriceDetailSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Price
+        fields = '__all__'
```
Same

KittenAstronaut commented on trading/serializers.py:
```diff
@@ -0,0 +1,53 @@
+            if data['entry_quantity'] > data['quantity']:
+                raise serializers.ValidationError("u cant sale stocks more than u have")
+        if data['order_type'] == 'Purchase':
+            if data['price'] < 10:
```
may be `data['order_type'] == 'Purchase' and data['price'] < 10:` instead of yours

KittenAstronaut commented on trading/serializers.py:
```diff
@@ -0,0 +1,53 @@
+class TradeSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Trade
+        fields = '__all__'
```
Same

KittenAstronaut commented on trading/serializers.py:
```diff
@@ -0,0 +1,53 @@
+class TradeDetailSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Trade
+        fields = '__all__'
```
Same

KittenAstronaut commented on user/managers.py:
```diff
@@ -0,0 +1,24 @@
+
+    @staticmethod
+    def create(user, currency, balance=0):
+        from user.models import Wallet
```
Why local imports?

KittenAstronaut commented on user/managers.py:
```diff
@@ -0,0 +1,24 @@
+
+    @staticmethod
+    def create(user, item):
+        from user.models import Watchlist
```
Why local imports?

KittenAstronaut commented on trading/services.py:
```diff
@@ -0,0 +1,113 @@
+        print(inventory_buyer)
+
+        inventory_seller = Inventory.objects.filter(item=seller_offer.item, user=seller_offer.user).first()
+        print(inventory_seller)
```
remove prints

KittenAstronaut commented on trading/services.py:
```diff
@@ -0,0 +1,113 @@
+from django.db import transaction
+
+
+class TradeService:
```
just a best practice. When you are using a method only inside of class, its name should start from _ for example: `_update_offer()`, `_update_inventory()`, etc.

KittenAstronaut commented on user/views.py:
```diff
@@ -0,0 +1,145 @@
+                     'watchlist': stats['item_in_watchlist_count'],
+                     'sold_offers': stats['sold_offers_count'],
+                     'bought_offers': stats['bought_offers_count']
+                     }
```
I don't see any reason to rename dict fields

KittenAstronaut commented on user/serializers.py:
```diff
@@ -0,0 +1,137 @@
+
+    def validate(self, data):
+
+        if data['currency'].is_not_deleted is False:
```
in `if`, `while` and another statements boolean should not to be compared with anything.
```python
if not data['currency'].is_not_deleted:
```

Commit:
Added docker-compose, fix dockerfile

KittenAstronaut commented on user/serializers.py:
```diff
@@ -0,0 +1,137 @@
+    def update(self, instance, validated_data):
+        instance.balance += validated_data.get('balance', 0)
+        instance.save(update_fields=('balance',))
+        return instance
```
You do not need there balance as a field, if you write it in Meta class

KittenAstronaut commented on Dockerfile:
```diff
@@ -0,0 +1,9 @@
+
+RUN pip install -r requirements.txt --no-cache-dir
+
+ENTRYPOINT [ "/bin/bash", "/opt/djangoTrading/entrypoint.sh" ]
```
Remake `Dockerfile` according to our yesterday conversation

KittenAstronaut commented on .dockerignore:
```diff
@@ -0,0 +1,5 @@
+*.sqlite3
+
+# static
+/djangoTrading/static/
```
`venv` are also not needed for docker

KittenAstronaut commented on pytest.ini:
```diff
@@ -0,0 +1,3 @@
+[pytest]
+DJANGO_SETTINGS_MODULE = djangoTrading.settings
+python_files = tests.py test_*.py *_tests.pyx
```
i think you are not needed in `*_tests.pyx`

KittenAstronaut commented on trading/enums.py:
```diff
@@ -0,0 +1,10 @@
+
+    @classmethod
+    def get_order_types(cls):
+        return [(tag.value,tag) for tag in cls]
```
For readability you have missed space between items in tuple

KittenAstronaut commented on requirements.txt:
```diff
@@ -0,0 +1,50 @@
+urllib3==1.26.6
+vine==5.0.0
+wcwidth==0.2.5
+django-environ==0.7.0
```
why are you have two django-environs?

KittenAstronaut replied:
and as i said above it's better to use python-dotenv

KittenAstronaut commented on djangoTrading/conftest.py:
```diff
@@ -0,0 +1,51 @@
+
+
+@pytest.fixture
+def data() -> Dict:
```
if you are using type hinting on return value it's better to assign a types of data for example `Dict[int, float]` where every key is int and every value is float.
If there's situation where key or value could be type of int and str, so `Dict[Union[int, str], float]`. key - int or str, value float.
If value could be anything so use `Dict[int, Any]`

KittenAstronaut replied:
For your case `Dict[str, Union[str, int]]`

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -68,16 +82,48 @@
+REDIS_HOST = env('REDIS_HOST')
+REDIS_PORT = env('REDIS_PORT')
+# CELERY_ALWAYS_EAGER = False
+CELERY_BROKER_URL = 'redis://' + REDIS_HOST + ':' + REDIS_PORT + '/0'
```
**DO NOT** use + operator on strings. Better use:
Format strings: 
```python
f'redis://{REDIS_HOST}:{REDIS_PORT}/0'
```
Format method:
```python
'redis://{}:{}/0'.format(REDIS_HOST, REDIS_PORT)
```

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -68,16 +82,48 @@
+# CELERY_ALWAYS_EAGER = False
+CELERY_BROKER_URL = 'redis://' + REDIS_HOST + ':' + REDIS_PORT + '/0'
+CELERY_BROKER_TRANSPORT_OPTIONS = {'visibility_timeout': 3600}  
+CELERY_RESULT_BACKEND = 'redis://' + REDIS_HOST + ':' + REDIS_PORT + '/0'
```
Same as above

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -68,16 +82,48 @@
     'default': {
-        'ENGINE': 'django.db.backends.sqlite3',
-        'NAME': BASE_DIR / 'db.sqlite3',
+        'ENGINE': 'django.db.backends.postgresql_psycopg2',
```
better to take out engine to `.env` too

KittenAstronaut commented on docker-compose.yaml:
```diff
@@ -0,0 +1,25 @@
+version: "3.3"
+
+services:
+  backend:
+    container_name: djangotrading_container
+    build: .
+    ports:
+      - '8000:8000'
+    depends_on:
+      - postgres
+      - redis
```
make docker-compose updates when you are updating code
using `volumes` as we talk yesterday

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -68,16 +82,48 @@
+    ),
+}
+
+# REDIS related settings
+# REDIS_HOST = '0.0.0.0'
+# REDIS_PORT = '6379'
+REDIS_HOST = env('REDIS_HOST')
+REDIS_PORT = env('REDIS_PORT')
+# CELERY_ALWAYS_EAGER = False
```
remove useless comments

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -11,21 +11,27 @@
 DEBUG = True
 
-ALLOWED_HOSTS = []
+ALLOWED_HOSTS = [
+    '*'
+]
```
allowed hosts should be also in `.env`
As usual you write in `.env` file something like:
```ini
ALLOWED_HOSTS="localhost 0.0.0.0"
```
Or any needed hosts and then in `settings.py`
```python
ALLOWED_HOSTS = os.environ.get('ALLOWED_HOSTS', 'localhost').split() # usual case
ALLOWED_HOSTS = env('ALLOWED_HOSTS').split() # your case
```

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -11,21 +11,27 @@
+SECRET_KEY = env('SECRET_KEY')
 
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
```
`DEBUG` variable is also goes to `.env`. For your case you can use example from documentation

KittenAstronaut commented on trading/services.py:
```diff
@@ -0,0 +1,113 @@
+            quantity=quantity
+        )
+
+    def making_trade(self):
```
making told that service currently makes trade. here we want to call it `make_trade`

KittenAstronaut commented on trading/services.py:
```diff
@@ -0,0 +1,113 @@
+                    is_active=True,
+                    order_type=OrderType.SALE.value,
+                    ).order_by('price').first()
+                if seller_offer is not None:
+                    quantity = min(buyer_offer.entry_quantity, seller_offer.entry_quantity)
+                    trade = Trade(
+                        item=buyer_offer.item,
+                        seller_offer=seller_offer,
+                        buyer_offer=buyer_offer,
+                        quantity=quantity,
+                        unit_price=buyer_offer.price
+                    )
+                    trade_list.append(trade)
+                    with transaction.atomic():
+                        self.update_all_fields(
+                            seller_offer=seller_offer,
+                            buyer_offer=buyer_offer,
+                            quantity=quantity
+                        )
+                else:
+                    break
```
i found another solution
```python
                if seller_offer is None:
                    break
                quantity = min(buyer_offer.entry_quantity, seller_offer.entry_quantity)
                trade = Trade(
                        item=buyer_offer.item,
                        seller_offer=seller_offer,
                        buyer_offer=buyer_offer,
                        quantity=quantity,
                        unit_price=buyer_offer.price
                    )
                trade_list.append(trade)
                with transaction.atomic():
                    self.update_all_fields(
                          seller_offer=seller_offer,
                          buyer_offer=buyer_offer,
                          quantity=quantity
                      )
```

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,95 @@
+        fields = ('code', 'name')
+        read_only_field = ('id',)
+
+    code = serializers.CharField(max_length=8)
+    name = serializers.CharField(max_length=128)
```
in some places you need to specify Field directly but it's for `SlugRelatedField` etc.
Here there is no reason to specify them

KittenAstronaut reviewed (changes requested):
Now you need:
* Make changes according to comments
* Improve code style using flake8
* Split serializers in actions
About last:
Best practice is to split serializers. One called something like `UpdateItemSerializer` is used by one action (`update` for this case`). If there's any serializers for one model that took same fields for different actions you can unite them and call `UpdateCreateItemSerializer`

Commit:
Added nginx and gunicorn conf, fix dockerfile, docker-compose. Some fixes in apps, codes tyle fixes with flake8

Commit:
refactored serializers, added rabbitmq for celery broker, redis uses for celery result backend, refactored docker-compose

Commit:
fix docker-compose , no all settings for worker and scheduler takes from settings.py

KittenAstronaut commented:
First of all. Keep time in commit messages permanent. Commonly it's present simple `add requirements.txt`, past simple `added requirements.txt`.
Sometimes projects use naming conventions. For example some of them firstly write what they hav done bugfix, feature, refactor.
For your project it will be looks like
`fix: change Dockerfile`
`feat: add RabbitMQ support for Celery`
`refactor: split serializers`

KittenAstronaut commented on configuration/nginx.conf:
```diff
@@ -0,0 +1,19 @@
+
+    location / {
+        proxy_pass http://backend:8000;
+    }
```
Most common situation. All calls to api goes to /api route. And all others will give static elements. So it will be `location /api`

KittenAstronaut commented on djangoTrading/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from __future__ import absolute_import, unicode_literals
```
You do not need `unicode_literals` when writing in Python 3
The same as `absolute_import`

KittenAstronaut commented on djangoTrading/celery.py:
```diff
@@ -0,0 +1,19 @@
+from __future__ import absolute_import, unicode_literals
```
Same

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -11,22 +11,26 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = 'django-insecure-0z=7ybt!dh85py0=id=r=78r#*6*c7y9cd2*gr)7q$*j^d2^bt'
+SECRET_KEY = str(os.getenv('SECRET_KEY'))
```
environmental variables are always strings

KittenAstronaut commented on djangoTrading/settings.py:
```diff
@@ -11,22 +11,26 @@
-DEBUG = True
-
-ALLOWED_HOSTS = []
+DEBUG = str(os.getenv('DEBUGMODE', True))
```
it looks strange. Maybe i've made a mistake was describes it to you. Even if it will be False, debug will True, because string is not empty.
The better way is:
.env file:
```ini
DEBUG_MODE=true
```
settings.py:
```python
DEBUG = os.environ.get('DEBUG_MODE', 'true').lower() == 'true'
```

KittenAstronaut commented on entrypoint.sh:
```diff
@@ -0,0 +1,40 @@
+  ./manage.py collectstatic
+
+  echo "Starting gunicorn"
+  gunicorn djangoTrading.wsgi \
+          --bind 0.0.0.0:8000 \
+          --reload \
+          --max-requests 100 \
+          --threads 2 \
+          --access-logfile -
```
I told you to split config into `gunicorn.conf.py`

KittenAstronaut commented on entrypoint.sh:
```diff
@@ -0,0 +1,40 @@
+  ;;
+beat)
+  start_beat
+  ;;
```
and also to split all entrypoints into different files

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,127 @@
+        if data['name'] in bad_words:
+            raise serializers.ValidationError("sorry, dont "
+                                              "use bad words")
+        if data['code'] not in data['code'].upper():
```
Also, there is another function:
```python
'O PURPLE FLAME O WHIRLING WHEEL OV LIFE APPEAR! MOVE! RISE!'.is_upper() #True
```

KittenAstronaut replied:
And again you compare strings with `in` operator

KittenAstronaut replied:
And also you can split it into different methods `validate__name` and others

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,127 @@
+        model = Item
+        fields = ('price', 'currency', 'details')
+        read_only_field = ('id', )
+    currency = CurrencyRetrieveSerializer()
```
And also first goes Serializers, after that goes class Meta

KittenAstronaut replied:
And commonly they are separated by break line

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,127 @@
+        model = Price
+        fields = ('price', 'currency', 'item', 'date')
+        read_only_field = ('id', )
+    currency = CurrencyRetrieveSerializer()
+    item = ItemListSerializer()
+    price = serializers.DecimalField(
+        validators=(positive_price_validator,), max_digits=20, decimal_places=2
+    )
```
same as above

KittenAstronaut commented on trading/models.py:
```diff
@@ -0,0 +1,65 @@
+        related_name='trade'
+    )
+    seller_offer = models.ForeignKey(Offer,
+                                     blank=True,
```
no needs to use `blank=True`

KittenAstronaut commented on trading/models.py:
```diff
@@ -0,0 +1,65 @@
+                                     related_name='seller_trade',
+                                     on_delete=models.CASCADE)
+    buyer_offer = models.ForeignKey(Offer,
+                                    blank=True,
```
same as above

KittenAstronaut commented on trading/models.py:
```diff
@@ -0,0 +1,65 @@
+    price = models.DecimalField(
+        max_digits=7,
+        decimal_places=2,
+        blank=True,
+        null=True
```
i think price cannot be null and blank

KittenAstronaut commented on trading/models.py:
```diff
@@ -0,0 +1,65 @@
+    item = models.ForeignKey(
+        Item,
+        on_delete=models.CASCADE,
+        related_name="offer"
```
Your Item model will have a set of offers, so i think you need to use plural `offers`. Like above in User

KittenAstronaut commented on trading/models.py:
```diff
@@ -0,0 +1,65 @@
+    """Information about a certain transaction"""
+    item = models.ForeignKey(
+        Item,
+        blank=True,
```
same as below

KittenAstronaut replied:
blank...

KittenAstronaut commented on trading/serializers.py:
```diff
@@ -0,0 +1,79 @@
+                raise serializers.ValidationError(
+                    "u cant sale stocks more than u have"
+                )
+        if data['order_type'] == 'Purchase' and data['price'] < 10:
```
import enum OrderType and use `data['order_type'] == OrderType.PURCHASE.value`
Same with `'Sale'` above

KittenAstronaut commented on trading/serializers.py:
```diff
@@ -0,0 +1,79 @@
+            raise serializers.ValidationError(
+                "u can enter only positive value"
+            )
+        if data['item'].currency.is_not_deleted is False:
```
`if not data['item'].currency.is_not_deleted`

KittenAstronaut commented on trading/views.py:
```diff
@@ -0,0 +1,99 @@
+    model = Offer
+    queryset = Offer.objects.all()
+    filterset_class = OfferUserFilter
+    permission_classes = (IsAuthenticated, BlackListPermission)
```
i think you do not need this declaration according to `permission_classes_by_action` below. And also first always goes variables and only after them goes methods. DO NOT MIX THEM

Commit:
refactor: 1) split dockerfiles and entrypoint.sh for services : Dockerfile_backend - run_backend.sh, Dockerfile_worker - run_worker.sh and Dockerfile_scheduler - run_scheduler.sh
2) added configuration for api and admin in nginx.conf
3) target Dockerfiles for services in docker-compose.yaml
4) split validators in serializers and replaced class Meta down
5) Deleted module __future__ from __init__.py and celery.py
6) Switched value for DEBUG in settings.py
Added: added settings for gunicorn in gunicorn.conf.py

Commit:
add back requirements.txt

KittenAstronaut commented on Dockerfile_sheduler:
```diff
@@ -0,0 +1,17 @@
+FROM python:3.8
+
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+ENV SERVICE=${run_service:-backend}
+ENV QUEUES=${queues:-celery}
+
+WORKDIR /opt/djangoTrading/
+
+RUN pip install --upgrade pip
+COPY run_scheduler.sh /opt/djangoTrading/
+COPY requirements.txt /opt/djangoTrading/
+RUN pip install -r requirements.txt --no-cache-dir
+
+COPY . /opt/djangoTrading/
+
+ENTRYPOINT [ "/bin/bash", "/opt/djangoTrading/run_scheduler.sh" ]
```
No, you have no need to separate Dockerfiles. Main aim is to save one Dockerfile and make it reusable for celery and django using [entrypoint command](https://docs.docker.com/compose/compose-file/compose-file-v3/#entrypoint) in docker-compose

KittenAstronaut commented on configuration/gunicorn.conf.py:
```diff
@@ -0,0 +1,9 @@
+bind = '0.0.0.0:8000'
+workers = 2
+worker_connections = 1000
+threads = 2
+timeout = 30
+keepalive = 32
+max_requests = 100
+user = 'nobody'
+accesslog = '-'
```
**FYI**, the default gunicorn setup uses
```python
import multiprocessing as mp
workers = mp.cpu_count() * 2 + 1
```
But it's NOT always true. It's depends on your server. Sometimes it's too much when operations on your server isn't cpu bound. Sometimes it's not enough.

**P.S.** You don't need to change that 'cause your laptop could fucked up.

KittenAstronaut commented on configuration/gunicorn.conf.py:
```diff
@@ -0,0 +1,9 @@
+bind = '0.0.0.0:8000'
```
**FYI**, I prefer to split it to
```python
host = '0.0.0.0'
port = 8000
bind = f'{host}:{port}'
```

KittenAstronaut commented on docker-compose.yaml:
```diff
@@ -0,0 +1,74 @@
+version: "3.3"
```
**FY**I, when we start a new project we prefer to use latest stable versions for docker-compose it's `3.9`

KittenAstronaut commented on item/serializers.py:
```diff
@@ -0,0 +1,148 @@
+
+    class Meta:
+        model = Currency
+        fields = ('name', 'code')
+        read_only_field = ('id',)
```
I'm not sure but it will be good if you'll specify all `read_only_fields` explicitly in fields
And also read_only_field**s** not read_only_field

KittenAstronaut replied:
Change it everywhere ctrl+shift+r will help you

KittenAstronaut commented on item/views.py:
```diff
@@ -0,0 +1,113 @@
+    def get_all_currency(self, request) -> Response:
+        """Allows to get all  'currency' in system,
+         available only for not banned users"""
+        queryset = Currency.objects.all()
```
For sure i cannot understand why are you using there `objects`

KittenAstronaut commented on item/views.py:
```diff
@@ -0,0 +1,113 @@
+    """A viewset for model Currency that provides `retrieve`,
+    `create`, update' and `list` actions."""
+    model = Currency
+    queryset = Currency.objects.filter_not_deleted()
```
I think it will be good idea to split it to another manager
```python
queryset = Currency.active_objects.all()
```
and also it's a good manner to use all in queryset initialization

KittenAstronaut commented on run_worker.sh:
```diff
@@ -0,0 +1,7 @@
+./manage.py migrate
+
+echo "Starting celery worker"
+celery -A djangoTrading.celery worker --loglevel=INFO --concurrency=1 -n worker2@%h
```
`2@%h`?

KittenAstronaut commented on trading/views.py:
```diff
@@ -0,0 +1,98 @@
+        return self.serializer_action_classes[self.action]
+
+    def get_permissions(self):
+        try:
+            return [permission() for permission
+                    in self.permission_classes_by_action[self.action]]
+        except KeyError:
+            return [permission() for permission in self.permission_classes]
```
Error handling takes a long time. So i think there will be great another option.
```python
permissions = self.permission_classes_by_action.get(self.action, self.permission_classes)
return [permission() for permission in permissions]
```

KittenAstronaut commented on user/models.py:
```diff
@@ -0,0 +1,85 @@
+    user = models.ForeignKey(
+        User,
+        on_delete=models.CASCADE,
+        related_name='watchlists'
+    )
+    item = models.ForeignKey(
+        Item,
+        blank=True,
+        null=True,
+        on_delete=models.CASCADE,
+        related_name='watch_list'
```
Try to keep same naming conventions for related names, for example plural and without underscores `watchlists`

KittenAstronaut commented on user/models.py:
```diff
@@ -0,0 +1,85 @@
+    )
+    item = models.ForeignKey(
+        Item,
+        blank=True,
```
and again blank...

KittenAstronaut commented on user/models.py:
```diff
@@ -0,0 +1,85 @@
+    )
+    item = models.ForeignKey(
+        Item,
+        blank=True,
```
blank...

KittenAstronaut commented on user/models.py:
```diff
@@ -0,0 +1,85 @@
+    )
+    currency = models.ForeignKey(
+        Currency,
+        blank=True,
```
blank..

KittenAstronaut commented on user/permissions.py:
```diff
@@ -0,0 +1,34 @@
+    def has_permission(self, request, view):
+        user = request.user
+        blacklist = BlackList.objects.filter(user=user).first()
+        if not blacklist:
+            return True
```
`return not blacklist`

KittenAstronaut commented on user/permissions.py:
```diff
@@ -0,0 +1,34 @@
+
+    def has_object_permission(self, request, view, obj):
+
+        if obj.balance > 0:
+            return False
+        else:
+            return True
```
`return obj.balance <= 0`

KittenAstronaut commented on user/serializers.py:
```diff
@@ -0,0 +1,199 @@
+        read_only_field = ('id',)
+
+    def create(self, validated_data):
+        return Wallet.objects.create(
+            user=validated_data['user'],
+            currency=validated_data['currency'],
+            balance=validated_data['balance']
```
**FYI**, i think for better readability in the future it will be great to unpack `validated_data` first
```python
user, currency, balance = (
    validated_data['user'],
    validated_data['currency'],
    validated_data['balance'],
)
return Wallet.object.create(user=user, currency=currency, balance=balance)
```

KittenAstronaut commented on user/views.py:
```diff
@@ -0,0 +1,165 @@
+        return self.serializer_action_classes[self.action]
+
+    def get_permissions(self):
+        try:
+            return [permission() for permission
+                    in self.permission_classes_by_action[self.action]]
+        except KeyError:
+            return [permission() for permission in self.permission_classes]
```
same

KittenAstronaut reviewed (approved):
You don't need to fix **FYI** comments it's just to adjust your understanding
Great job

Commit:
add ignore static root in  .gitignore

Commit:
refactored: 1) deleted blank=True from models where it's not need
2)renamed read_only_field in serializers to read_only_fields
3)in user.permissions logic made more easy
4)in views refactored permission for actions
5)in docker-compose.yaml get back 1 Dockerfile, added entrypoint for every service
6)in Dockerfile added permissions for all entrypoints
7) in run_worker.sh and run_scheduler.sh added latency for RabbitMQ
8)in run_backend.sh added latency for Postgresql
Added: 1)added new manager for Currency model
2)added one Dockerfile for all services

KittenAstronaut committed:
Merge pull request #3 from awertynaxis/easyfix

refactored: 1) deleted blank=True from models where it's not need
